libs/ha/ha_common_libs_k8s.py

```diff
@@ -1845,3 +1846,116 @@ def dnld_obj_verify_chcksm(self, s3_test_obj, bucket, obj, download_path, upload
+        return True, pass_put_ver if not background else output.put((True, pass_put_ver))
+
+    @staticmethod
+    def parallel_get_object(event, s3_ver_obj, bkt_name: str, obj_name: str, ver_etag: list,
+                            output=None, **kwargs):
+        """
+        Function to GET a version of an object in parallel or non parallel.
+        :param event: event to intimate thread about main thread operations
+        :param s3_ver_obj: S3VersioningTestLib instance
+        :param bkt_name: Target bucket for GET Object with VersionId call.
+        :param obj_name: Target key for GET Object with VersionId call.
+        :param ver_etag: Target list of dictionary of uploaded versions to be GET
+        :param output: Output queue in which results should be put
+        :return: Tuple (bool, list)
+        """
+        background = kwargs.get("background", False)
+        fail_get_ver = []
+        pass_get_ver = []
+        for v_etag in ver_etag:
+            v_id = list(v_etag.keys())[0]
+            etag = list(v_etag.values())[0]
+            if v_id == "null":
+                continue
+            try:
+                get_resp = s3_ver_obj.get_object_version(bucket=bkt_name, key=obj_name,
+                                                         version_id=v_id)
+                if get_resp[1]["VersionId"] != v_id or get_resp[1]["ETag"] != etag:
+                    failed = {v_id: [get_resp[1]["VersionId"], get_resp[1]["ETag"]]}
+                    fail_get_ver.append(failed)
+                else:
+                    pass_get_ver.append(v_etag)
+            except CTException as error:
+                LOGGER.exception("Error in %s: %s", HAK8s.put_get_delete.__name__, error)
+                if event.is_set():
+                    continue
+                fail_get_ver.append(v_etag)
+        if fail_get_ver:
+            return False, fail_get_ver if not background else output.put((False, fail_get_ver))
+        return True, pass_get_ver if not background else output.put((True, pass_get_ver))
+
+    @staticmethod
+    def list_verify_version(s3_ver_obj, bucket_name: str, expected_versions: dict):
+        """
+        Function to List all the versions of bucket objects and verify the output with expected
+        version etag values
+        :param s3_ver_obj: S3VersioningTestLib instance
+        :param bucket_name: Target bucket for List Object version
+        :param expected_versions: Target list of dictionary of uploaded versions to be List and
+        verify
+        :return: Tuple (bool, str)
+        """
+        resp = s3_ver_obj.list_object_versions(bucket_name)
+        v_etag = []
+        list_version = resp[1]["Versions"]
+        for v_e in list_version:
+            versions_dict = {v_e["VersionId"]: v_e["ETag"]}
+            v_etag.append(versions_dict)
+        for v_e in v_etag:
+            if v_e not in expected_versions:
```

Comment:
From where we will get expected_versions?

Reply:
> From where we will get expected_versions?

This will be passed from the steps we followed and tested in test. 
We are keeping track of all the object version till this function gets called.